wfe2/wfe.go

```diff
@@ -531,12 +535,186 @@ func (wfe *WebFrontEndImpl) regHoldsAuthorizations(ctx context.Context, regID in
+		if prob != nil {
+			wfe.sendError(response, logEvent, prob, nil)
+			return
+		}
+	default:
+		// Otherwise the request is malformed
+		wfe.sendError(response, logEvent,
+			probs.Malformed("Malformed JWS, no KeyID or embedded JWK"), nil)
+		return
+	}
+
+	// Read the revoke request from the JWS payload
+	type RevokeRequest struct {
+		CertificateDER core.JSONBuffer    `json:"certificate"`
+		Reason         *revocation.Reason `json:"reason"`
+	}
+	var revokeRequest RevokeRequest
+	if err := json.Unmarshal(jwsBody, &revokeRequest); err != nil {
+		wfe.sendError(response, logEvent, probs.Malformed("Unable to JSON parse revoke request"), err)
+		return
+	}
+
+	// Parse the provided certificate
+	providedCert, err := x509.ParseCertificate(revokeRequest.CertificateDER)
+	if err != nil {
+		wfe.sendError(response, logEvent, probs.Malformed("Unable to parse certificate DER"), err)
+		return
+	}
+
+	// Compute and record the serial number of the provided certificate
+	serial := core.SerialToString(providedCert.SerialNumber)
+	logEvent.Extra["ProvidedCertificateSerial"] = serial
+
+	// Lookup the certificate by the serial. If the certificate wasn't found, or
+	// it wasn't a byte-for-byte match to the certificate requested for
+	// revocation, return an error
+	cert, err := wfe.SA.GetCertificate(ctx, serial)
+	if err != nil || !bytes.Equal(cert.DER, revokeRequest.CertificateDER) {
+		wfe.sendError(response, logEvent, probs.NotFound("No such certificate"), err)
+		return
+	}
+
+	// Parse the certificate into memory
+	parsedCertificate, err := x509.ParseCertificate(cert.DER)
+	if err != nil {
+		// InternalServerError because cert.DER came from our own DB.
+		wfe.sendError(response, logEvent, probs.ServerInternal("invalid parse of stored certificate"), err)
+		return
+	}
+	logEvent.Extra["RetrievedCertificateSerial"] = core.SerialToString(parsedCertificate.SerialNumber)
+	logEvent.Extra["RetrievedCertificateDNSNames"] = parsedCertificate.DNSNames
+
+	// Check the certificate status for the provided certificate to see if it is
+	// already revoked
+	certStatus, err := wfe.SA.GetCertificateStatus(ctx, serial)
+	if err != nil {
+		wfe.sendError(response, logEvent, probs.NotFound("Certificate status not yet available"), err)
+		return
+	}
+	logEvent.Extra["CertificateStatus"] = certStatus.Status
+
+	if certStatus.Status == core.OCSPStatusRevoked {
+		wfe.sendError(response, logEvent, probs.Conflict("Certificate already revoked"), nil)
+		return
+	}
+
+	switch authType {
+	// If the request was authenticated by key ID then the associated account
+	// must have authorizations for all of the names in the certificate that is being
+	// revoked. Otherwise it is not authorized for this revocation.
+	case embeddedKeyID:
+		// ServerInternal because this should not happen. We get acct from
+		// `validPOSTForAccount` when authType == embeddedKeyID and it returns an error if
+		// the account doesn't exist
+		if acct == nil {
```

Comment:
This should be moved up to before we set `acct.ID` after calling `validPOSTForAccount`.